### DISCUSSION TITLE
Is there a way to prevent an expression from being used in an update clause?

### DISCUSSION THREAD

**Discussion by dnaaun** [diesel-rs/diesel#3606] [Q&A]
I'm working on writing a Diesel DSL for window functions (initially postgres, and only the functions I'm currently interested in, but the plan is to clean it up and either make a PR or release it as a separate crate), and I discovered that Postgres forbids window functions in update statements. Can/should I express that in Diesel?

**Comment by weiznich** [comment]
First of all great to hear that someone started to work on window functions. That's something I would be interested in adding to diesel itself.

That written: I think there is currently no such trait that allows you to restrict that special expressions couldn't appear in update statements. If possible we might want to add such a restriction. If that's not possible adding this restriction is something that we should consider for a potential diesel 3.0 release.

I personally would prefer cutting a 2.1 release first (that mainly means that I must solve #3566 + write some more documentation). After that I'm happy to spend some time helping with a concrete design here.

**Reply by dnaaun** [reply]
Sorry for the delay.

We can work on the design at your leisure!

I'm not sure how to go about the design / implement / review cycle, so I'll just outline what I've done below, and let you decide how we should go about it (ie, what I should do first, etc).

1. Implemented a DSL (not in the diesel crate) to use the syntax of [postgres aggregate expressions](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-AGGREGATES), excluding the syntax for _ordered set aggregate functions_ (ie, `WITHIN GORUP`). Example usage is

```rust
array_agg(id).order_by(orig_text.desc()).filter(not(id.eq_any(excluded)))
```

2. Implemented a DSL (not in the diesel crate) to use [postgres window function calls](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS) with the variants of the syntax that avoid having a `window_name`. Also, the DSL doesn't currently express the `frame_clause` postgres expression type, but that should be straightforward to add. Example usage is:

```rust
cume_dist().over(partition_by(recipe_source).order_by(aggregate_rating_value.desc()))
```

In both of the above, the DSL is designed in such a way that one can easily add specific aggregate functions/window functions by writing a struct, implementing a trait for that struct, and adding a constructor function that returns that struct (all of which can be hidden away in a macro call, I imagine).

Functionality wise, here are the TODOs, I think:

1. Implement `ordered set aggregate functions`, which I think should be straightforward.
2. The current DSL for aggregate expressions doesn't forbid an invalid expression: `ARRAY_AGG(DISTINCT *)` (ie, `DISTINCT` with `*`), fix that.
3. Postgres docs say "any built-in or user-defined ordinary aggregate (i.e., not ordered-set or hypothetical-set aggregates) can be used as a window function", the DSLs currently don't express that.
4. Extend the DSL to allow specifying a `frame_clause` for window function calls.
5. Express the invariant that aggregate/window expressions cannot be nested inside other aggregate/window expressions.
6. Express the invariant that window function calls can only be used in `SELECT` and `ORDER BY` clauses (it turns out `UPDATE` is not the only thing they are not allowed to appear in ).

I think it should be ok to avoid supporting _named windows_, since Postgres guarantees that "When multiple window functions are used, all the window functions having syntactically equivalent PARTITION BY and ORDER BY clauses in their window definitions are guaranteed to be evaluated in a single pass over the data.". And so one should just be able to assign to a variable a closure producing a window definition in Rust, and use that in place of where one would use named windows.

**Reply by weiznich** [reply]
There is no need to be sorry about a delay here. It's great to see some work in this area.
 
I think I personally would like to break the work down into smaller steps, so that you try to implement a minimal thing first and we add features later on. That makes it easier to review. In terms of your todo list (which looks great) this means that we probably want at least one PR per step. Feel free to open these PR's early on to get feedback on the design or to use them to ask questions about how to implement certain steps. In that case: Be prepared that we likely need some iterations to get there.

The first step is to perform some research which part of that DSL is postgres specific and which part might be useful for other backends (including custom ones) as well and which invariants exists.

The proposed DSL looks ok, although we might want to have some more discussion around the `OVER(_)` part of window functions as the current proposal requires a freestanding `partition_by` function but that part is optional (including all other clauses there). 

> In both of the above, the DSL is designed in such a way that one can easily add specific aggregate functions/window functions by writing a struct, implementing a trait for that struct, and adding a constructor function that returns that struct (all of which can be hidden away in a macro call, I imagine).

This part already exists via the `sql_function!` macro. It is required to extend that macro to generate the necessary code for window functions as that's also used by diesel internally for functions like `max` and `min`.

> I think it should be ok to avoid supporting named windows, since Postgres guarantees that "When multiple window functions are used, all the window functions having syntactically equivalent PARTITION BY and ORDER BY clauses in their window definitions are guaranteed to be evaluated in a single pass over the data."

That's fine.
